### PR TITLE
parser: add EXPORT statement syntax parsing

### PR DESCRIPTION
**Pull request by dt**
Implementation to follow, but the export statement will allow exporting
arbitrary query results in CSV format to a designated storage location.

Release note: none.

### CODE REVIEW

**Comment by dianasaur323** [comment]
:lgtm: although just my two cents, I think the syntax of CSV () SELECT is a bit confusing. Although INSERT doesn't use a key word, it does use VALUES to separate INSERT INTO () VALUES ()...

---







*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24728#-:-L9ukytjDQyTFE75_91s:b5ho585)*

**Comment by knz** [comment]
@dianasaur323 this proposal both supports `EXPORT CSV () VALUES ()` and `EXPORT CSV() SELECT ..`, exactly like INSERT does.

File: docs/generated/sql/bnf/stmt_block.bnf [hunk 1]
```diff
@@ -83,6 +84,9 @@ explain_stmt ::=
 	'EXPLAIN' explainable_stmt
 	| 'EXPLAIN' '(' explain_option_list ')' explainable_stmt
 
+export_stmt ::=
+	'EXPORT' import_data_format '(' string_or_placeholder opt_with_options ')' select_stmt
```
Comment (knz) [inline docs/generated/sql/bnf/stmt_block.bnf]: So what was the rationale for not having INTO here? (as in `EXPORT INTO CSV ...`)

File: pkg/sql/parser/sql.y [hunk 2]
```diff
@@ -1595,6 +1597,26 @@ import_stmt:
...
+//
+// Formats:
+//    CSV
+//
+// Options:
+//    comma = '...'          [CSV-specific]
+//
+// %SeeAlso: SELECT
+export_stmt:
+  EXPORT import_data_format '(' string_or_placeholder opt_with_options ')' select_stmt
```
Comment (knz) [inline pkg/sql/parser/sql.y]: the WITH outside of the parentheses?

**Comment by dt** [comment]
I had to move the opts inside the `()` to avoid a conflict though... but that doesn't match IMPORT anymore.
OTOH, import takes a _list_ of files, whereas export doesn't -- it takes a base path, and puts some number of files there. not sure if that matters as far as wrapping in `()`. also, fwiw, `BACKUP` and `RESTORE` don't use `()`.

---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24728#-:-L9ulE8rEl1WIImfBDGD:b-g6bkbk)*

File: docs/generated/sql/bnf/stmt_block.bnf (hunk 1 above)
Comment (dt) [reply docs/generated/sql/bnf/stmt_block.bnf]: nothing in particular, just didn't seem to need it? does it read better with it?

File: pkg/sql/parser/sql.y (hunk 2 above)
Comment (dt) [reply pkg/sql/parser/sql.y]: whoops, just commented top-level about that, but i got a conflict with it immediately next to `select_stmt`. I'm also not completely sure about the `()` for that matter.

File: docs/generated/sql/bnf/stmt_block.bnf (hunk 1 above)
Comment (mjibson) [reply docs/generated/sql/bnf/stmt_block.bnf]: We already have `BACKUP ... TO` and `RESTORE ... FROM`. `EXPORT INTO` reads better to my ears.

**Review by madelynnblue — COMMENTED** [review]
Ok now I'm ready to bikeshed. I prefer `EXPORT INTO CSV 'target' WITH opts FROM select_stmt`. This adds the `INTO` and `FROM` keywords. It appears that we either have to have to FROM keyword or deal with the very wide `( target WITH opts )` thing in parens to avoid the conflict. I think the parens being around that much stuff looks real bad.

**Comment by knz** [comment]
+1 on Matt's suggestions.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by dt** [comment]
I also like matt's suggested syntax. Revising with that unless anyone else chimes in.

**Comment by knz** [comment]
For the record the ambiguity would not exist if the ccl stuff was using USING for options, as is standard for all the other SQL statements, instead of WITH which is standard for CTEs. It's really sticking out.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by knz** [comment]
To followup on my previous point you now have

 EXPORT INTO .... WITH comma = XXX FROM WITH foo = yyy SELECT

Which is properly defined (and its semantics are desirable!), but reads ugly and is super hard to explain.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by dt** [comment]
@knz those `WITH kv_opts` things are a hold-over from when were trying to iterate on a branch for all enterprise code, that made frequent sql.y changes for changing options expensive. I think now that we're past that, it'd make sense to just promote all our options to the grammar and do away with the string-ly typed map.

.. BUT i don't want to do that here. If you like this plan, let's do it en masse, to all the users of kv_opts. And we have until 2.1 to tweak and break `EXPORT` syntax, so I don't want to let that refactor be a blocker here.

**Comment by knz** [comment]
Yes I agree. My remarks were for the record, not an immediate invitation for change.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by dt** [comment]
TFTRs. Updated with matt's suggested syntax, RFAL.

**Review by madelynnblue — APPROVED** [review]

File: pkg/sql/parser/sql.y [hunk 3]
```diff
@@ -1595,6 +1597,26 @@ import_stmt:
...
+// %Help: EXPORT - export data to file in a distributed manner
+// %Category: CCL
+// %Text:
+// EXPORT <format> (<datafile> [WITH <option> [= value] [,...]]) <query>
+//
+// Formats:
+//    CSV
+//
+// Options:
+//    comma = '...'          [CSV-specific]
```
Comment (mjibson) [inline pkg/sql/parser/sql.y]: `delimiter`

Comment (dt) [reply pkg/sql/parser/sql.y]: hmm. I was thinking symmetry with `IMPORT`.

Comment (mjibson) [reply pkg/sql/parser/sql.y]: https://github.com/cockroachdb/cockroach/blob/cf21dc74090b69b364dfd8630843cc9af33be7fb/pkg/ccl/importccl/csv.go#L55

Comment (dt) [reply pkg/sql/parser/sql.y]: huh, okay then.
Done.

**Comment by mjibson** [comment]
LGTM

**Comment by dt** [comment]
bors r=mjibson

**Pull request merged**